### ISSUE TITLE
Return type for functionRunner

### ISSUE THREAD

**Issue by BenJohnCarson** [willmorgan/stub-azure-function-context#43]
Currently the [`functionRunner`](https://github.com/willmorgan/stub-azure-function-context/blob/develop/lib/function-runner.ts#L13) has a return type of `Promise<any>`. 
This results in a few `typescript-eslint` errors:

Assigning result : 
![Screenshot 2023-06-15 at 10 57 23](https://github.com/willmorgan/stub-azure-function-context/assets/21341022/23c71ef7-0a13-45f8-be12-3f26e398b19f)

Trying to type with `Promise<Context>` : 
![Screenshot 2023-06-15 at 11 01 02](https://github.com/willmorgan/stub-azure-function-context/assets/21341022/4f2a7987-4b8e-4f7d-b6bc-ce6b0ff6df77)

For now we've got around this by adding an overload to our project:

```
import * as stubAzureFunctionContext from 'stub-azure-function-context';

declare module 'stub-azure-function-context' {
  export function functionRunner(
    azFunction: AzureFunction,
    bindingDefinitions?: stubAzureFunctionContext.BindingDefinition[] | string,
    bindingData?: Record<string, stubAzureFunctionContext.Binding>,
    augmentContext?: stubAzureFunctionContext.AugmentContextCallback,
  ): Promise<Context>;
}
```

It'd be ideal if the type were narrowed in the package itself. Is there a reason it's currently `any`? Is `Context` a sufficient type?

**Comment by dhensby** [comment]
The return type of the function runner is not a `Context` so we shouldn't be setting it so at all.

It is `any` (though I suppose could be `unknown`) because the function can return... anything.

If you see the [docs](https://learn.microsoft.com/en-us/azure/developer/javascript/how-to/develop-serverless-apps?tabs=v4-ts#a-simple-javascript-function-for-http-requests), it can be an HTTP response, or a void function, etc. I'm not aware of any time the function would (or should) return the context.

I'm not really clear on what the linting error is there, perhaps because the screen shots don't show enough context, but I can't tell why or where you'd be wanting to assert the type is a `Context`. Do you have some more complete code to show us?

**Comment by dhensby** [comment]
OK - right, my confusion here is that the `functionRunner` returns the context if the function itself doesn't actually return anything.

So if your underlying function is not returning a value, you'll get the context back, but if it does return something then you get *that* back. So we can't just declare the return type as a `Context` because the return type is a `Context` in some scenarios, but not in any where the function has a return value.

**Comment by BenJohnCarson** [comment]
Ah ok interesting. In that case would allowing passing a generic type for the return of the underlying function be a possible option? Perhaps having it default to `unkown` if not supplied. Something like :
```typescript
export async function functionRunner<T = unkown>(
    azFunction: AzureFunction, 
    bindingDefinitions: BindingDefinition[] | string = [], 
    bindingData: Record<string, Binding> = {}, 
    augmentContext?: AugmentContextCallback
): Promise<T> { ...
```

**Comment by dhensby** [comment]
It could be generic, or maybe even using the return type of `AzureFunction` itself?

**Comment by dhensby** [comment]
OK - there doesn't seem to be a way to infer the return type, which would be a bit nicer than forcing the use of a generic.

I've added it as a generic in #44, but I would prefer to add a conditional return type if we could evaluate if the azure function was a void function or not...

**Comment by BenJohnCarson** [comment]
Yeh that'd be ideal. I'm not sure it's possible though because of the `AzureFunction` type that comes from the `@azure/functions` package ([here](https://github.com/Azure/azure-functions-nodejs-library/blob/2b2177c7f5eee98d2451df0e59e69490391752f5/types/index.d.ts#L26)). There it's always typed as `Promise<any> | void` so unless the consumer passed a function to `functionRunner` that used a custom type other than `AzureFunction` it'd always evaluate to the same type.

**Comment by dhensby** [comment]
Well, if a function is declared to return `Promise<string>` that is a valid `AzureFunction` which has been narrowed, so it should be possible in *theory* to obtain that narrowed return type... though I've not been able to 😅

**Comment by dhensby** [comment]
OK - I think I've managed to get it working so it will infer the return type successfully.

Do you think you could install the branch that #44 is based on and see if that resolves your problems?

**Comment by BenJohnCarson** [comment]
Unfortunately it still defaults to a return type of `Promise<any>`.

<img width="746" alt="Screenshot 2023-06-19 at 15 38 11" src="https://github.com/willmorgan/stub-azure-function-context/assets/21341022/b3e8e7e1-4bd1-48d3-970b-2ce61a4249e1">

It does work if I tweak it to this (adding the `any | void`): 
```typescript
export declare function functionRunner<T extends AzureFunction = AzureFunction>(
  azFunction: T,
  bindingDefinitions?: BindingDefinition[] | string,
  bindingData?: Record<string, Binding>,
  augmentContext?: AugmentContextCallback,
): Promise<Awaited<ReturnType<T>> extends any | void ? Context : ReturnType<T>>;
```
But then I don't think it's possible to get it to have a return type of anything other than Context as `extends any | void` I think always evaluates to true. I do think the `AzureFunction` type is the problem here as it doesn't allow determining the return type

**Comment by BenJohnCarson** [comment]
It may just be that the way our team has written our functions is a little niche and perhaps not quite the intended approach. e.g. our functions will always have a return type of `Promise<void>` because we just set the response to `context.res`:
```typescript
export const httpTrigger: AzureFunction = async (context): Promise<void> => {
  context.res = {
    status: 200,
    body: {
      status: 'Healthy',
    },
  };
};
```

**Comment by dhensby** [comment]
When I check it locally the typings seem to be resolving even for your example signature, I've even turned on the `@typescript-eslint/no-unsafe-assignment` rule and it passes for your example.

![image](https://github.com/willmorgan/stub-azure-function-context/assets/563596/47d65ef1-82af-4084-98d1-9546711b5347)

So it doesn't like the default `any` type that is inferred from the stub, but it's happy with the more specific typings if the function is a void function or returns a narrower type.

**Comment by BenJohnCarson** [comment]
Yeh looks like it works perfectly if I change my function to not use the `AzureFunction` type and instead explicitly type it. For some reason it ignores/overrides the declared return type when typing the function with `AzureFunction`:
```typescript
// triggers eslint rule in test, functionRunner return type any
export const httpTrigger: AzureFunction = async (context): Promise<void> => {}

// doesn't trigger eslint rule, functionRunner return type Context
export const httpTrigger = async (context: Context): Promise<void> => {}
```
I'm happy with this solution if you are!

**Comment by dhensby** [comment]
fixed in #44 and released in [v2.2.1](https://github.com/willmorgan/stub-azure-function-context/releases/tag/v2.2.1)
